### PR TITLE
Swift: Taint models for NSString

### PR DESCRIPTION
**Pull request by geoffw0**
Taint models for `NSString`, plus `NSMutableString` and a partial model of `NSObject`.  These are fairly thorough but there are some test cases we don't yet model, e.g. flow into closures, and there are probably a few rows in the model that aren't explicitly tested.

### CODE REVIEW

**Review by AlexDenisov — DISMISSED** [review]

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsString.qll [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        ";NSString;true;components(separatedBy:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;trimmingCharacters(in:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;substring(from:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;substring(with:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;substring(to:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;folding(options:locale:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;applyingTransform(_:reverse:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;replacingOccurrences(of:with:);;;Argument[-1];ReturnValue;taint",
+        ";NSString;true;replacingOccurrences(of:with:);;;Argument[1];ReturnValue;taint",
+        ";NSString;true;replacingOccurrences(of:with:);;;Argument[-1];ReturnValue;taint",
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsString.qll:91]: This seems like a bug, @geoffw0?

[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsString.qll:91]: It does.  I'll check if it should have been something else, or whether I can just delete it...

[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsString.qll]: Done (it was supposed to be a different variant of the same named method).

**Review by AlexDenisov — APPROVED** [review]

**Comment by geoffw0** [comment]
I don't know what's gone wrong with the PR checks (and DCA run), but I need to fix a merge conflict anyway (from https://github.com/github/codeql/pull/12226), so we'll see if that makes our problems go away...

**Comment by MathiasVP** [comment]
> I don't know what's gone wrong with the PR checks (and DCA run), but I need to fix a merge conflict anyway (from #12226), so we'll see if that makes our problems go away...

Looking at the PR check failure I see:
> 2023-02-17T10:19:03.5321505Z swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsObject.qll would change by autoformatting.

And the DCA run looks like it failed to build the CLI.

**Comment by geoffw0** [comment]
> 2023-02-17T10:19:03.5321505Z swift/ql/lib/codeql/swift/frameworks/StandardLibrary/NsObject.qll would change by autoformatting.

Thanks, I didn't see that in the log.  Will fix and start a new DCA run.

**Comment by geoffw0** [comment]
The last two DCA runs failed during the CLI step:
```
unzip:  cannot find or open codeql-*-extractors-part-win64.zip, codeql-*-extractors-part-win64.zip.zip or codeql-*-extractors-part-win64.zip.ZIP.
```
Any suggestions? (besides waiting a bit, merging main, and re-running DCA)

**Comment by geoffw0** [comment]
^ looks like its a problem affecting all DCA for Swift right now.  Fix coming apparently.

**Comment by geoffw0** [comment]
Merged in main and a new DCA run started...

**Comment by geoffw0** [comment]
DCA LGTM.

**Comment by geoffw0** [comment]
Fixed test failures (caused by the change to inline expectations on `main`).

**Review by MathiasVP — DISMISSED** [review]
The QL changes LGTM. I don't fully understand the CI failures, though.

**Comment by geoffw0** [comment]
I think some built-in is probably missing on Linux but present on Mac, and the test assumes it's there.  But the following log output is not super helpful when it's from CI rather than an interactive session:
```
Extraction command .../qltest.sh failed with status 1. Use --show-extractor-output to see its output.
```

**Comment by geoffw0** [comment]
I've managed to debug this in a codespace, the issue was:
```
nsstring.swift:97:38: error: cannot find type 'AutoreleasingUnsafeMutablePointer' in scope
```

**Pull request merged**